[CI] Add a make + windows github action job.

NlightNFotis opened the pull request:
This is a draft (for now) PR that adds a `windows` * `make` Github Actions
job. This is using the already existing `cmake` one as a template.

<!---
Thank you for your contribution. Please make sure your pull request fulfils all of the below requirements. If you cannot currently tick all the boxes, but would still like to create a PR, then add the label "work in progress" and assign the PR to yourself.
--->

- [ ] Each commit message has a non-empty body, explaining why the change was made.
- [ ] Methods or procedures I have added are documented, following the guidelines provided in CODING_STANDARD.md.
- [ ] The feature or user visible behaviour I have added or modified has been documented in the User Guide in doc/cprover-manual/
- [ ] Regression or unit tests are included, or existing tests cover the modified code (in this case I have detailed which ones those are in the commit message).
- [ ] My commit message includes data points confirming performance improvements (if claimed).
- [ ] My PR is restricted to a single feature or bugfix.
- [ ] White-space or formatting changes outside the feature-related changed lines are in commits of their own.

<!---
See, e.g., https://chris.beams.io/posts/git-commit/ for general guidelines on commit messages.

If you have created commits mixing multiple features and/or unrelated white-space changes, use a sequence involving git reset and git add -p to fix this.
--->

NlightNFotis committed:
Add a make + windows CI configuration for github actions.

NlightNFotis committed:
Change make commons to invoke win_flex and win_bison.

When a user installs winflexbison on Windows, by default
the binaries are named win_flex and win_bison. This commit
aligns our configuration with the environment default.

NlightNFotis committed:
Ignore regression tests marked winbug on windows make builds.

This includes tests under:
  * regression/cpp
  * regression/cbmc-cpp
  * regression/systemc

TGWDB commented on regression/cbmc-cpp/Makefile:
```diff
@@ -4,7 +4,7 @@ include ../../src/config.inc
 
 ifeq ($(BUILD_ENV_),MSVC)
-	excluded_tests = -X gcc-only
+	excluded_tests = -X gcc-only - X winbug
```
Extra space in "- X"?

NlightNFotis replied:
Good catch!

thomasspriggs commented on src/common:
```diff
@@ -146,11 +146,11 @@ ifeq ($(origin CXX),default)
 endif
 ifeq ($(origin YACC),default)
-  YACC   = bison -y
+  YACC   = win_bison -y
```
:pick: I am guessing this is required due to the names used by the versions of flex/bison we are using in the new job. But can you add the rationale to the commit message body so A) I don't have to guess B) Anyone who starts seeing local failures due to using different versions, knows there was a reason for doing this. Not just to ruin their day ;)

kroening replied:
May I suggest to simply override this variable when calling make. This way, local builds won't be broken.

NlightNFotis replied:
It's worth noting here that when someone installs `winflexbison` (https://github.com/lexxmark/winflexbison) the default binaries are called `win_flex` and `win_bison`. It takes an extra manual step to rename these to `flex` or `bison`

tautschnig replied:
Is this the only place that people could possibly source flex/bison for Windows from? Putting it differently, I don't recall hearing from anyone that they were unhappy with our Makefiles for Windows, which either means that no one is using them or existing users have a working set-up. So I'm with @kroening: set `YACC='win_bison -y'` and `LEX=win_flex` on the command-line.

thomasspriggs replied:
There are many ways to get Flex / Bison on Windows. If a user has a build environment based on Cygwin or MSYS2 for example then the packages for those environments may still be using the original names. I'd need to check to know for sure, but I don't have a Windows machine on hand.

NlightNFotis replied:
@tautschnig Are we sure we want to do that? Just for clarification, there are 3 potential configurations for building on windows (that I can see in `src/common`), and they are Cygwin, MSVC and MinGW based. For them:

1. The `MinGW` configuration is inheriting `flex` and `bison` binary names from the default configuration,
2. The `Cygwin` configuration also sets them explicitly to `flex` and `bison`, and
3. The `MSVC` configuration is the one I changed to address `win_flex` and `win_bison`, which are the default binaries a user should end up with if he installs `winflexbison` (which is the default way a programmer would acquire these to use them with Visual Studio today, to my understanding).

So, I guess my question is: Does this really break anything? Alternative build environments on Windows should not be affected in any capacity (if a user is using Cygwin or MinGW) and the MSVC one I can argue that was misconfigured (or at least stale by current standards).

thomasspriggs replied:
I have now taken another look at the rest of this file. I have now seen that there are other sets of configuration in this file for cygwin and mingw. Therefore I am less concerned about explicitly changing this for Visual Studio only. I am happy for this change to be merged as is.

tautschnig replied:
> So, I guess my question is: Does this really break anything? [...] and the MSVC one I can argue that was misconfigured [...]

I simply did not know whether `winflexbison` was the (single?) most common approach. If you're confident that's the case, then, sure, let's go with it. My concern was breaking existing user's set-ups, but if it seems unlikely that anyone was (successfully) using this configuration, then there shouldn't be a problem. Well, if there is, people will tell us and we can revert :-)

thomasspriggs commented on .gitattributes:
```diff
@@ -7,5 +7,5 @@
 *.css text
 *.inc text
-test.desc text
+*.desc text eol=lf
```
:pick: Can you add the explanation about the line endings causing problems with `test.pl` to the commit message body please?

thomasspriggs commented on regression/cpp/Makefile:
```diff
@@ -10,7 +10,7 @@ else
 
 ifeq ($(BUILD_ENV_),MSVC)
-	excluded_tests = -X gcc-only
+	excluded_tests = -X gcc-only -X winbug
```
I am kinda surprised this wasn't failing on the old codebuild job. Obviously not a blocking comment, but I wonder why that was passing.

NlightNFotis replied:
I have had a thought about this, and I've arrived at two theories regarding this, but I can't test them now:

1) The `makefiles` were misconfigured before, not including whole folders from `regression/` so it might just be that some tests were skipped by accident.
2) Even though the `_BUILD_ENV` was also specified to be `MSVC`, it ended up being overridden to `Cygwin` because of the build environment, which was calling `g++` which does not throw compiler errors for these tests (`cl.exe` does though, which is what `clcache` is invoking).

thomasspriggs replied:
Hmm. I thought this was supposed to be a visual studio build. I.E. not using g++ at all. Still not a blocking comment though, as we need to get this merged rather than continuing to have no tests of Windows + make.

NlightNFotis replied:
This *is* a Visual Studio build only - I've gone to great lengths to ensure that the only compiler used is `cl.exe` through `clcache`.

I'm talking about the previous code build job this is replicating. The reason I said I'm assuming it ended up using `g++` is a weird explicit setup of the path to Cygwin tools before it called `make`.

thomasspriggs replied:
Ah, now I understand. :+1: It makes sense that the tests could have been failing silently before due to incorrect setup.

thomasspriggs reviewed (commented):
One general nit pick here is that your commit messages titles should not end in a full stop and your commit message bodies should explain why you are making the change. See - [Chis Beams full stop](https://chris.beams.io/posts/git-commit/#end) and [Chris Beams why not how](https://chris.beams.io/posts/git-commit/#why-not-how).

Nit picks aside, congratulations on finally getting this working.

thomasspriggs reviewed (approved):
LGTM.

chrisr-diffblue reviewed (approved):
It seems like the contentious part of the PR only affects VS builds, and not _all_ windows builds - which perhaps was a bit of context missing from the original PR/commit history. Happy for this to go in.

tautschnig commented on .gitattributes:
```diff
@@ -7,5 +7,5 @@
 *.css text
 *.inc text
-test.desc text
+*.desc text eol=lf
```
Can we fix `test.pl` instead? I've read rumours that Strawberry Perl isn't doing a good job adapting to the OS, so maybe we should just accept that there are perl implementations that don't play ball. I believe all that it takes is replacing `chomp` in line 79 of `test.pl` by some use of `s/\R$//`.

tautschnig replied:
More concretely: Replacing `chomp @data;` by `s/\R$// for @data;` should do the trick.

NlightNFotis replied:
I'm happy with this if this works. I will wait for CI and if this is working, I am going to merge it.

NlightNFotis committed:
Make test.pl capable of handling CRLF.

Previously we had problems with test.pl choking on test.desc
file input on windows as a result of the test.pl files being
checkout out by git with CRLF instead of LF as line endings.
